### ISSUE TITLE
PLEASE TEST: SpaceFM 1.0 Alpha-Testing Has Begun

### ISSUE THREAD

**Issue by IgnorantGuru** [IgnorantGuru/spacefm#478]
The next major version (SpaceFM 1.0) is available for early alpha testing.  This version finishes up some work completed last year by myself and OmegaPhil on various kinds of handlers, includes BwackNinja's feature addition to set a transparent background (now without restarting or command-line options), and should soon include video thumbnail support.  This version is still under light development so this should be used mostly for testing, and for contributing default handlers for your favorite filesystems, protocols, and archive functions.  **Please help test this version so the release will be more stable.**

SpaceFM 1.0 (currently 0.9.9+) addresses some shortcomings in SpaceFM 0.9:

* Only a single mount or unmount command was available for dealing with all filesystem types.
* The few network and other fuse filesystems that SpaceFM handled were usually mounted by udevil.  But being a suid program, udevil is not the best choice for this, and more filesystems were desired.
* Additional protocols required writing a script that handled all protocols, and unmounting of protocols were channeled through only the one unmount command.
* Only a limited number of hard-coded archive types were supported by SpaceFM's built-in archive facility, and options were limited.
* Bypassing MIME types for opening files using custom methods was limited to creating custom menu items, which was difficult to manage.

To address these, SpaceFM 1.0 includes a new configuration window for creating and changing handlers for devices, protocols, archive files, and regular files.  The device and protocol handlers add potential support for many more filesystems, and the archive handlers (developed mostly by long-time SpaceFM contributor OmegaPhil) allow virtually any archive types to be handled in unique ways.  File handlers allow you to bypass the MIME system for file types.

Other changes in this version include:

* The spacefm <a href="http://ignorantguru.github.io/spacefm/spacefm-manual-hand-en.html#invocation-file">command line can now open devices</a> (eg /dev/sdd1) and protocols (eg ftp://)
* Devices (eg /dev/sdd1/) can now be entered in the path bar or a bookmark
* A change detection blacklist has been added which allows the user to determine which filesystems should have changes ignored (to improve GUI performance).  This setting will also disable thumbnail loading soon.
* New added menu item File|Open URL (replaces unneeded Save Session item)
* Show|Display Name can use %n to show device major:minor number
* The file list in GTK2 now selects the next file after the current file is deleted or renamed.
* Upcoming video thumbnail support to be added soon based on some code contributions by Justin Jacobs ([dorkster@github](http://dorkster.github.com/))

To install SpaceFM 1.0 for testing, first backup your SpaceFM 0.9 config folder for insurance:  [**UPDATE: Please see [updated install instructions](https://github.com/IgnorantGuru/spacefm/issues/478#issuecomment-91026515)**]
~~~~
cp -a ~/.config/spacefm ~/.config/spacefm-0.9.4-backup
~~~~
Then install the hand branch:
~~~~
# Download & Extract
mkdir /tmp/spacefm-hand && cd /tmp/spacefm-hand
wget -O spacefm-hand.tar.gz https://github.com/IgnorantGuru/spacefm/tarball/hand
tar xzf spacefm-hand.tar.gz
cd IgnorantGuru-spacefm-*

# Build & Install
./configure --prefix=/usr  # and optionally: --with-gtk3 see ./configure --help
make -s
sudo make install
sudo update-mime-database /usr/share/mime > /dev/null
sudo update-desktop-database -q
sudo gtk-update-icon-cache -q -t -f /usr/share/icons/hicolor
sudo gtk-update-icon-cache -q -t -f /usr/share/icons/Faenza

# Remove Temporary Files
cd / && rm -rf /tmp/spacefm-hand
~~~~

You can copy and paste all of those commands into your terminal as a set.  Because this branch is still under some development, it is best to update it every few days by repeating the above commands, and [check issues related to the hand branch](https://github.com/IgnorantGuru/spacefm/milestones/hand) and [see latest changes](https://github.com/IgnorantGuru/spacefm/commits/hand).

When first running 0.9.9, any custom mount and unmount command you have configured will be imported into the new Default device handler.  Any protocol handler script you had set will be imported as a protocol handler, overriding the new default handlers.  Any options you used for creating archives in 0.9 will need to be reconfigured in the new Archive Handlers dialog.  Also running 0.9 with your updated configuration should not be a problem.

**New default handlers are welcome!**  If you use additional filesystems or protocols, please consider submitting handlers for them to improve SpaceFM's out of the box compatibility.  Please keep default handlers simple and basic - usually a few lines will do.  You can submit them as an issue on this tracker or email them to me.  Remember to include whitelist, blacklist, and all related commands.  (There is not currently a way to import or export handlers.)

The new version is documented in the user's manual in the new [Handlers](http://ignorantguru.github.io/spacefm/spacefm-manual-hand-en.html#handlers) section, and also [Opening Files, URLs, and Devices](http://ignorantguru.github.io/spacefm/spacefm-manual-hand-en.html#invocation-file) and elsewhere.

Transifex has an updated translation file for this branch - [please help update translations](https://github.com/IgnorantGuru/spacefm/wiki/Translate) using their online editor.

Also, donations to help support this project are once again [being accepted](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=FF9Y4Q845CBCN) - thanks.

Minor issues may simply be brought to attention by posting a comment below, or you can open a separate issue.  Thanks for testing and contributing.

**Comment by VastOne** [comment]
Configured and installed 0.9.9 and happily testing. I have also instructed all VSIDO users to assist if they can

Will report on anything I find, valid or not... ;)  

Good to see you again IgnorantGuru

**Comment by IgnorantGuru** [comment]
VastOne, thanks as always.  Glad to see you guys made it through the service interruption.

**Comment by IgnorantGuru** [comment]
Hand branch now includes video thumbnail support. This means that to build it, libffmpegthumbnailer-dev must be installed, or use configure option --disable-video-thumbnails

**Comment by mati75** [comment]
I can push this version to Debian experimental repository, but I need two days for this.

**Comment by RaveYoda** [comment]
Loving the added transparent bg option so I can use nitrogen or xwinwrap as my bg setter!! Also, THANK YOU for finally adding video thumbnailing!! It really is a great addition. May I suggest a seperate icon size manager for the desktop? Having it tied to the Fm's internal size isn't the most optimal as I'd like to have large icons in the file mngr itself with small icons on the desktop itself.

**Comment by Vladimir-csp** [comment]
Transparent desktop works great!
Although, I think that this option does not logically belong to wallpaper mode list.

**Comment by fabertawe** [comment]
Many thanks for all the fixes and continued work :) Hand's running great here so far (Arch) and I use it extensively.

**Comment by VastOne** [comment]
Not sure if it is an issue or is intentional but the build made no mention of needing fuseiso now as a dependency to mount ISO's

Once fuseiso was installed (debian) ISO's mounted again in 0.9.9+

**Comment by IgnorantGuru** [comment]
VastOne, currently the default Mount ISO (a file handler) uses only fuseiso.  I may rewrite that to use either fuseiso or udevil, whichever is installed.  fuseiso was a little easier to deal with (though its use of its own mtab file wasn't).  By default, that handler is not a default action - you choose it from the menu.  Thus README was updated with 'suggests fuseiso', as a runtime dependency.

In general with handlers if you get a 'command not found', that's the cue that the handler requires a program.  But in this case it can also be done with a udevil mount iso handler.

**Comment by IgnorantGuru** [comment]
A few changes as of https://github.com/IgnorantGuru/spacefm/commit/a0db27f1be90f6367ea23d4a664c92db34f78850 in hand:

* SpaceFM now uses a somewhat more inclusive algorithm when determining what non-block device, mounted filesystems (like fuse or nfs) to show in the Devices list.  A filesystem will be shown if it is associated with a protocol handler via fstype or mtab_fs=fstype.  Otherwise it will only be shown if it is NOT of type: devpts proc fusectl pstore sysfs tmpfs devtmpfs  (let me know if you think another belongs in that list).  Also, the mount point directory must be readable by the user or it will not be shown.  What do you think of these new defaults?  This mostly affects what is shown when Settings|Show|Mounted Other is enabled.

* File Handlers are now run as synchronous tasks, will popup and error on non-zero exist status, and will be shown in the task manager.  To prevent this, run apps in the handler with an ampersand suffix per the manual.  This provides better error reporting.  Drawback includes the handler writer must include an ampersand or the task will stay until the app closes.

* The default file handler Mount ISO will now use either fuseiso or udevil, and it will popup errors.  The device handler default have also been changed.  (Use the Defaults or  All Defaults buttons to pull in new defaults.)

* I will shortly add some additional default protocol handlers from Vladimir-csp

**Comment by Vladimir-csp** [comment]
When mounting something with protocol handlers, the process is referred to as "Network mount". Probably this should be replaced with some neutral phrasing. Not every handler is network-related

**Comment by kanyck** [comment]
Thank you guys for this awasome (master)piece of software! Been looking forward to this release!
Is there an ebuild available? I'd like to participate in testing but try to avoid installing anything bypassing portage...

**Comment by OmegaPhil** [comment]
Some mounted filesystem types that should probably be not be shown by default:

cgroup (mounted at '/sys/fs/cgroup/systemd')
binfmt_misc (/proc/sys/fs/binfmt_misc)
rpc_pipefs (/run/rpc_pipefs)

Shame that an aufs volume doesn't have a label to display.

**Comment by IgnorantGuru** [comment]
@OmegaPhil > Shame that an aufs volume doesn't have a label to display.

Why is it is a shame?  If you think it's useful to see, we could hack in a command to populate the label with something useful, providing it is executable by the user.  I made aufs hidden but that could be removed, or any fstype that has a protocol handler is shown, overriding the blacklist.

If you're interested, hack for aufs label could go in [here](https://github.com/IgnorantGuru/spacefm/blob/d2bbb356a4597e906a054cb1e1d3761071d8bbc6/src/vfs/vfs-volume-nohal.c#L2984)

**Comment by OmegaPhil** [comment]
The aufs filesystem is basically the only permanent filesystem I have that doesn't support labels, so it appears blank - however that probably just meant I needed to review the display name setting - changing it to '%t: %l' is good enough, since I only have one aufs volume on the system.

**Comment by mati75** [comment]
I pushed autobuild hand branch for Ubuntu: https://code.launchpad.net/~mati75/+archive/ubuntu/spacefm-unstable

**Comment by OmegaPhil** [comment]
Currently I have thumbnails disabled (Preferences -> General tab, File List), but it appears that when downloading a file (and watching the directory), when SpaceFM deals with the change notification, thumbnails enable but then switch off again - when the files finished downloading, thumbnails remained until I refreshed the directory.

**Comment by OmegaPhil** [comment]
Ah I see you added aufs to the automatic hidden list - please can this be allowed? Just updated.

**Comment by OmegaPhil** [comment]
Not sure if this will be considered a bug, but I'm also getting lots of spam on stderr/out:

===============================================

[matroska,webm @ 0x7f4ad81a2000] Unknown entry 0xF0
[matroska,webm @ 0x7f4ad81a2000] Unknown entry 0xB2

===============================================

As in, more than 5000 lines since SpaceFM was started. I guess the previous bug is causing thumbnail code to run, and it doesn't like something in one of the directories?

**Comment by IgnorantGuru** [comment]
@OmegaPhil Yes that thumbnailer api function writes a bunch of debug stuff to stdout/err (although I didn't see it do it when thumbnailing is disabled, will take a look).  I would love to shut it up but I don't see any debug level passed into the function.  Is there a way to close stdout/err to one function?

**Comment by OmegaPhil** [comment]
I haven't heard of anything to kill stdout/stderr so specifically.

**Comment by meuwe** [comment]
When I first ran spacefm and I clicked the hard drive image to view devices, a path for fuse was added, this I've not seen in 0.9.4 before.

See screenshot below;

http://i.imgur.com/fiBwn94.jpg

**Comment by IgnorantGuru** [comment]
@meuwe Thanks for the info.  With the latest update that gvfsd-fuse shouldn't be shown.  Appearance is completely up to the theme.  I'll take a look at the video thumbnail size options.

@OmegaPhil aufs remains hidden by the updated logic unless you add a protocol handler which includes it as the protocol or mtab_fs=aufs.  aufs is probably something which should not normally be shown but your input is welcome.

The new logic is that if there is a protocol handler for a non-block device's mounted fs type, then it will be shown (if the mount point is user-readable).  Otherwise it will be shown if it is not blacklisted (devpts proc fusectl pstore sysfs tmpfs devtmpfs ramfs aufs overlayfs cgroup binfmt_misc rpc_pipefs fuse.gvfsd-fuse) AND it is mounted within ~/.cache, /media/ or /run/media/.  Any non-blacklisted fs type that begins with "fuse." will also be shown even if mounted elsewhere.

This logic gets kind of involved because it's somewhat guesswork but this behavior is more like 0.9.4 handled it, hiding some types.  Now protocol handlers can override anything that would otherwise be hidden.

**Comment by meuwe** [comment]
@IgnorantGuru I'm assuming for new updates we are still going to download from?

```wget -O spacefm-hand.tar.gz https://github.com/IgnorantGuru/spacefm/tarball/hand```

I looked back up at the post and I didn't notice anything telling us when a new update is available. 

Where do we check for updates?

**Comment by IgnorantGuru** [comment]
Yes hand branch is most current.  You can see the latest commits [here](https://github.com/IgnorantGuru/spacefm/commits/hand).  If a commit is related to an issue usually you'll see a reference in the issue when a commit is made, such as the https://github.com/IgnorantGuru/spacefm/commit/84bae9fa242dd1976a0ea1361f227ce26b5ff019 ref above.  Mostly just polishing a few things at this point, and any comprehensive testing is appreciated.

**Comment by meuwe** [comment]
ok thanks

**Comment by Vladimir-csp** [comment]
When next file is selected after deleting, hitting enter does nothing. File is selected, focus seems to be in the file list, but item can not be activated until I move selection with arrows or mouse.

**Comment by cusco-travel-services** [comment]
Will this version support ptp only cameras?

**Comment by Vladimir-csp** [comment]
@cusco-travel-services use gphoto:// handler for this if gphotofs is installed

**Comment by cusco-travel-services** [comment]
Hi Vladimir,

Can the gphoto:// handler be configured to automatically mount ptp only cameras?

If so, how do I do it?

**Comment by Vladimir-csp** [comment]
I doubt that, unless IgnorantGuru adds some mechanism of discovering and listing available devices into handlers.

**Comment by cusco-travel-services** [comment]
Thanks Valdimir and I hope I.G. can manage to include such a feature. It would be very useful.

**Comment by IgnorantGuru** [comment]
@meuwe
> There is no way to have this in git so we can git clone then git pull to keep updated on it?

Yes, you can clone the hand branch if you like.  My generic instructions are for people not familiar with git.

**Comment by IgnorantGuru** [comment]
@Vladimir-csp Keyboard focus of next item should be corrected in 4e6c1cd

**Comment by IgnorantGuru** [comment]
@Vladimir-csp @cusco-travel-services
> I doubt that, unless IgnorantGuru adds some mechanism of discovering and listing available devices into handlers.

If you want automatic mounts on camera connection, you'll probably want to set this up at the system level.  eg when that device is inserted, run <code>spacefm gphoto://</code>

For example, if you're using PolicyKit you can probably create a rule in there.  Also, anytime a device is added, [evt-device](http://ignorantguru.github.io/spacefm/spacefm-manual-en.html#sockets-events-device) event occurs, and you can add an event handler for that.  This would remove the need for PolicyKit, etc.

You could also try forcing automount by adding the device in Devices|Settings|Auto Mount|Mount Volumes.  It must be seen as a block device by udev for that to work, however.

As for SpaceFM detecting such a device and automounting it, SpaceFM uses a set of criteria to determine if an added device should be automounted.  If someone wants to research the criteria for ptp devices this could be changed.

**Comment by Vladimir-csp** [comment]
The problem is - ptp and mtp devices are not block devices and are not directly mountable.
Here is my log and udev events when connecting android device in mtp mode:
log:
```
Apr 01 19:02:42 hostname kernel: usb 4-1.3.1.4: new high-speed USB device number 42 using ehci-pci
Apr 01 19:02:42 hostname kernel: usb 4-1.3.1.4: New USB device found, idVendor=04e8, idProduct=6860
Apr 01 19:02:42 hostname kernel: usb 4-1.3.1.4: New USB device strings: Mfr=1, Product=2, SerialNumber=3
Apr 01 19:02:42 hostname kernel: usb 4-1.3.1.4: Product: SAMSUNG_Android
Apr 01 19:02:42 hostname kernel: usb 4-1.3.1.4: Manufacturer: SAMSUNG
Apr 01 19:02:42 hostname kernel: usb 4-1.3.1.4: SerialNumber: XXXXXXXX
Apr 01 19:02:42 hostname colord[7936]: Device added: sysfs-SAMSUNG-SAMSUNG_Android
Apr 01 19:02:43 hostname colord[7936]: Device added: sysfs-(null)
```
Udev monitor:
```
KERNEL[1116062.942653] add      /devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.3/4-1.3.1/4-1.3.1.4 (usb)
KERNEL[1116062.943027] add      /devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.3/4-1.3.1/4-1.3.1.4/4-1.3.1.4:1.0 (usb)
UDEV  [1116062.968360] add      /devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.3/4-1.3.1/4-1.3.1.4 (usb)
UDEV  [1116063.981951] add      /devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.3/4-1.3.1/4-1.3.1.4/4-1.3.1.4:1.0 (usb)
```

When connected it as ptp device:
```
Apr 01 19:07:21 hostname kernel: usb 4-1.3.1.4: new high-speed USB device number 44 using ehci-pci
Apr 01 19:07:21 hostname kernel: usb 4-1.3.1.4: New USB device found, idVendor=04e8, idProduct=6865
Apr 01 19:07:21 hostname kernel: usb 4-1.3.1.4: New USB device strings: Mfr=1, Product=2, SerialNumber=3
Apr 01 19:07:21 hostname kernel: usb 4-1.3.1.4: Product: SAMSUNG_Android
Apr 01 19:07:21 hostname kernel: usb 4-1.3.1.4: Manufacturer: SAMSUNG
Apr 01 19:07:21 hostname kernel: usb 4-1.3.1.4: SerialNumber: XXXXXXXX
Apr 01 19:07:21 hostname mtp-probe[11771]: checking bus 4, device 44: "/sys/devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.3/4-1.3.1/4-1.3.1.4"
Apr 01 19:07:21 hostname mtp-probe[11771]: bus: 4, device: 44 was not an MTP device
Apr 01 19:07:21 hostname colord[7936]: Device added: sysfs-SAMSUNG-SAMSUNG_Android
Apr 01 19:07:21 hostname colord[7936]: Device added: sysfs-(null)
```
```
KERNEL[1116385.766426] add      /devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.3/4-1.3.1/4-1.3.1.4 (usb)
KERNEL[1116385.766776] add      /devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.3/4-1.3.1/4-1.3.1.4/4-1.3.1.4:1.0 (usb)
UDEV  [1116385.795978] add      /devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.3/4-1.3.1/4-1.3.1.4 (usb)
UDEV  [1116385.799580] add      /devices/pci0000:00/0000:00:1d.0/usb4/4-1/4-1.3/4-1.3.1/4-1.3.1.4/4-1.3.1.4:1.0 (usb)

```

**Comment by IgnorantGuru** [comment]
@Vladimir-csp SpaceFM currently automounts only block devices, and udev only provides device info on block devices.  So I think SpaceFM would have trouble getting any info on that device using its current methods.

Not sure if evt-device will react to non-block, probably not.  You can try setting View|Events|Device to:

    spacefm -g --label "%e %f %v" &

Then insert the ptp device and see if you're notified.  If you are, you could modify that command to mount conditionally if it's the ptp device.

Overall, this sounds like a job for PolicyKit or a similar daemon, which could then run spacefm to mount it.

**Comment by Vladimir-csp** [comment]
Selecting next file after delete now fully works.
Another selection issues: 
- select renamed file.
- when reloading directory (F5) selection should stay on the same item if possible, but:
 - if you freshly enter some dir and select something, then reload, selection drops.
 - when you return to parent dir, selection is restored to the dir you've just been to. Select something else, reload - selection returns to previous dir.

Nothing happens when ptp or mtp device is connected. In fact, nothing happens even if block device is connected! SpaceFM outputs to stdout: ```udev added:   /dev/sdc```, but script for event is not executed, no label shown.

**Comment by Vladimir-csp** [comment]
Unsaved handler warnings look good!
Although I still find current button placement in this dialog unpertinent.

**Comment by IgnorantGuru** [comment]
@Vladimir-csp
> Nothing happens when ptp or mtp device is connected. In fact, nothing happens even if block device is connected! SpaceFM outputs to stdout: udev added: /dev/sdc, but script for event is not executed, no label shown.

I can't reproduce that.  You might try for device event handler:

    echo EVENT %e %f %v &

If you run in a terminal, you should see it executed on device added/removed/changed events.  If that's not working please open an issue, hopefully narrowing down what is triggering it in your case.  You're reporting some unusual problems recently (eg socket not working) so you may want to test a default installation on a generic setup if you've made unusual setup changes.

**Comment by Vladimir-csp** [comment]
I've figured it out: The device event handler does not work if devices side panel was not open since SpaceFM start. Once devices panel is opened at least once, handler starts working correctly. Should I open another issue?

Anyway, no luck with ptp/mtp devices.

**Comment by IgnorantGuru** [comment]
@Vladimir-csp
> The device event handler does not work if devices side panel was not open since SpaceFM start.

Thanks that has been fixed in b8b0c7e

**Comment by IgnorantGuru** [comment]
Two new socket commands have been added to the latest hand branch (a2a417e), to mount and unmount devices and URLs.  See [manual](http://ignorantguru.github.io/spacefm/spacefm-manual-hand-en.html#run-task-mount)

**Comment by IgnorantGuru** [comment]
@OmegaPhil
> Currently I have thumbnails disabled (Preferences -> General tab, File List), but it appears that when downloading a file (and watching the directory), when SpaceFM deals with the change notification, thumbnails enable but then switch off again - when the files finished downloading, thumbnails remained until I refreshed the directory.

This should be fixed in bac8aff.  And I still see no way to quiet the stderr output from ffmpeg.  Closing stderr in that thread closes it in all threads.

**Comment by IgnorantGuru** [comment]
@meuwe
> Also it would be nice if the video thumbnails were at the same size as the folder icons. Right now folder icons are much larger in appearance. Even at 96x96 you can barely make out the video thumbnails, they're to small.

> Here's a screenshot showing the difference in sizes between the folders and video thumbnails;

> http://i.imgur.com/GUy9dMO.jpg

The video thumbnails are appearing at the correct size, maximum 96x96.  The vertical height of a video thumbnail depends on the aspect ratio of the video.  If the video is 16:9 widescreen, then it will appear shorter than a 4:3 video.  In your screenshot the thumbnails have the same width as the folder icons.  The thumbnails are shorter due to the aspect ratio.  They can't get any larger without being wider.  So everything is working correctly here.  Only thing you could do is request that spacefm support a larger Large icon size - or you can add that request to #94.

**Comment by IgnorantGuru** [comment]
@OmegaPhil
> And I still see no way to quiet the stderr output from ffmpeg. Closing stderr in that thread closes it in all threads.

I did find a way to alias and close stderr, then reopen it:
~~~~
                int fd = dup(fileno(stderr));
                close(fileno( stderr));
                
                video_thumbnailer_generate_thumbnail_to_file( video_thumb, file_path, thumbnail_file );
                
                dup2(fd, fileno(stderr));
                close(fd);
~~~~

This does work.  However, the main problem is that SpaceFM is multithreaded, so any number of threads or external programs could be writing to its stderr while the thumbnail loading thread is running, so I think it would create missing segments in the output.

Also, I'm not sure of the impact on performance to close and open stderr on every thumbnail.

**Comment by IgnorantGuru** [comment]
**New Max Icon Sizes**
bd6c8c5 allows icon sizes (small and large) up to 256x256.  Detailed list view uses the small icon size for thumbnails, so this allows for very large thumbnails in detailed view as well as icon view (which uses the Large icon size).

Also, side pane and task list icon sizes are based on the Small icon size, but are now limited to 32x32.  This allow you to choose a very big Small icon size for the list view without having ridiculously large icons in the side panes.  There should probably be a separate setting for this icon size, but this provided a quick way to extend the icon sizes.

**Comment by OmegaPhil** [comment]
Confirmed with the latest code that I don't see thumbnails enabling anymore with a changing file. 

Unrelated but now I'm looking at stderr/out, whenever you launch an Info dialog for a file or directory, then close it, Glib complains about 'removing a source':

(spacefm:1305): GLib-CRITICAL **: Source ID 261 was not found when attempting to remove it

Happens every time. This happens in the current stable too, not sure if its worth bothering with this minor problem.

**Comment by Vladimir-csp** [comment]
#480 comes to mind. There should be separate controls for icon size in main pane (icon mode), main pane (list/detailed modes), side panes, toolbars. This would avoid any confusion and hidden limits.

**Comment by Vladimir-csp** [comment]
Separate file size limit for thumbnailing for images/video/other would be handy too. Because different amount of resources is needed to thumbnail different types of content.

**Comment by Vladimir-csp** [comment]
![256scr](https://cloud.githubusercontent.com/assets/4061903/6997918/576b9e1e-dbd7-11e4-8697-35a4b2292dba.png)

**Comment by OmegaPhil** [comment]
Ah, cheers.

**Comment by IgnorantGuru** [comment]
@OmegaPhil
> GLib-CRITICAL **: Source ID 261 was not found

lol this is a bug that dates back to PCManFM 0.5.  Must be a recent change in glib's verbosity caused the warning to appear.  This has been fixed in 7dbe393.

Wrapping should now be fixed in a719c25.

The thumbnails were indeed 128x128 max.  Now they will increase to 256x256 if a size larger than 128 is selected.  However, old thumbnails will not be recreated unless you delete them from ~/.thumbnails, otherwise they will only be scaled.

@Vladimir-csp
Videos do not obey the Pic Size prefs limit.  More adjustable icon sizes is of course useful, but that's a more involved project - not as simple as adding settings as the code wasn't designed for 4 panels, let alone individual icon sizes.  For this release just increasing the max sizes.

**Comment by RaveYoda** [comment]
@mati75 I just saw your post for the autobuild branch. Thanks for posting it! Saves me some time in getting the lattest updates for testing. =]

**Comment by IgnorantGuru** [comment]
dd6739d adds a new option View|Style|Large Icons for Detailed and Compact list styles.  If enabled, the list will use the large icon/thumbnails.  By assigning a key shortcut you can switch back and forth easily.  This is also a [four-state setting](http://ignorantguru.github.io/spacefm/spacefm-manual-en.html#gui-pan-mem), which means that SpaceFM will remember the Large Icons setting for each panel for each combination of panel layouts.

This should remove the need to use big Small icon sizes.  Also, the max side pane icon size has been raised to 48.  Any larger settings for Small Icon size in prefs will not be reflected in the side panes.  (Any thoughts on this?  Should it be higher, or should there be no limit?)

Please test the Large Icons setting as thoroughly as possible.

**Comment by Vladimir-csp** [comment]
I doubt that side pane icons would ever need to be larger than 24, no previews there.

Here is a screenshot: everything seems to be fine, except two things: "+" icon in devices pane and bookmark icons are smaller than other icons in side panes.
![liscr](https://cloud.githubusercontent.com/assets/4061903/7000454/d88a895c-dc2b-11e4-9c59-3df46a158cfc.png)

**Comment by Vladimir-csp** [comment]
With all recent changes to icons I propose adding a view toolbar. Shortcuts are a good thing, but if user tends to use mouse, navigating deep menu hierarchy is a pain.
View toolbar would solve this. It could contain view mode selector, +/- buttons to change current icons size (the demand exists, several users asked me where to find them), thumbnailer switch, may be something else. The controls are already in SpaceFM, I suppose it would be relatively easy to place them on a toolbar.

**Comment by IgnorantGuru** [comment]
On TVs (SpaceFM can be used as a media center), 32 or 36 on the side panes can be useful.  Also, 4K displays are now in use.

Only reason I can think of for why your side panes have varying icon sizes is that if your icon theme doesn't support a size, it will use the smaller size.  Not all themes have the larger icons.  On my theme they're all the same size icons in the side panes.

As for the view toolbar, I plan to add a socket command for changing the Large Icons setting, so one could make a plugin for the toolbar.  Most panel settings have socket commands, so there is no need for me to hardcode such tools - they can be done as plugins.  Also note that there is a side toolbar that could be used for these view plugins.

I've considered changing the thumbnail on/off setting to a per panel setting.  I think that's pretty simple.  And a socket command could go with that as well.  If it's not added to this release you might want to open a request.  Not sure how much demand there is for it.  My main hesitation is adding another rarely-used setting to the menus.

**Comment by OmegaPhil** [comment]
Just tested the changes, thanks, confirm things work here - it takes fucktonnes of RAM (690MB for max thumbnail size of one directory of 1,790 images) but looks good :)

**Comment by fabertawe** [comment]
The way I personally use larger thumbnails is when traversing/editing folders of photos, so being able to change the size of icons (in icon mode) per panel (rather than "Preferences -> Large Icons" globally) would be perfect.

**Comment by IgnorantGuru** [comment]
d32d0cb adds a "large_icons" socket property to programmatically change the View|Style|Large Icons setting.

For example, <b>to add a toolbar button that toggles Large Icons:</b>
In one of the toolbar config menus, right-click and select New|Command.  Enter title "Large Icons".  For command enter:

    spacefm -s set large_icons %a

On the Options tab, select Style|Checkbox.  Click OK.

**Comment by IgnorantGuru** [comment]
@fabertawe
> The way I personally use larger thumbnails is when traversing/editing folders of photos, so being able to change the size of icons (in icon mode) per panel (rather than "Preferences -> Large Icons" globally) would be perfect.

That is how the new Large Icons setting works, per panel, which is perhaps what you meant.  It just lets you choose between small and large, and the specific sizes of those are set in Prefs.

**Comment by IgnorantGuru** [comment]
ffmpegthumbnailer devs have responded positively to [my request](https://code.google.com/p/ffmpegthumbnailer/issues/detail?id=117) for less noise on stderr.

**Comment by fabertawe** [comment]
@IgnorantGuru 
> That is how the new Large Icons setting works, per panel, which is perhaps what you meant. It just lets you choose between small and large, and the specific sizes of those are set in Prefs.

Unless I'm missing something obvious then it's not what I meant. I don't want to use a small icon setting above 22x22 and I already use the large icon setting (36x36) for view style "Icons". I would love to be able to change this from 36x36 to 96x96 on the fly for individual panels.

As it is I have a key shortcut for Preferences and change it from there (globally). It's not any great hassle to do it this way anyway, it just seems clumsy. Thanks.

**Comment by IgnorantGuru** [comment]
The hand branch has now been merged into next, and will be deleted shortly.  Please update your scripts to use these install instructions for the next branch:

~~~~
# Download & Extract
mkdir /tmp/spacefm-next && cd /tmp/spacefm-next
wget -O spacefm-next.tar.gz https://github.com/IgnorantGuru/spacefm/tarball/next
tar xzf spacefm-next.tar.gz
cd IgnorantGuru-spacefm-*

# Build & Install
./configure --prefix=/usr  # and optionally: --with-gtk3 see ./configure --help
make -s
sudo make install
sudo update-mime-database /usr/share/mime > /dev/null
sudo update-desktop-database -q
sudo gtk-update-icon-cache -q -t -f /usr/share/icons/hicolor
sudo gtk-update-icon-cache -q -t -f /usr/share/icons/Faenza

# Remove Temporary Files
cd / && rm -rf /tmp/spacefm-next
~~~~

The next branch represents a beta version pending release for 1.0.0, so any final testing is appreciated.

**Comment by humpity** [comment]
"The file list in GTK2 now selects the next file after the current file is deleted or renamed."
After renaming or deleting a file, next selection works, but hitting enter does nothing.
Moving away with arrow keys and then going back solves the problem.

**Comment by IgnorantGuru** [comment]
@humpity
> After renaming or deleting a file, next selection works, but hitting enter does nothing.
Moving away with arrow keys and then going back solves the problem.

Thanks, there was still a problem with Compact and Icon list styles.  This should now be corrected in next branch (4dc2e35).

**Comment by humpity** [comment]
Related to issue #444, the  v0.9.4-1 version resized percentage-wise (e.g 2-pane 50/50 width would resize to
same proportions. But alpha version does not do this anymore, leaving divider in same spot.

**Comment by IgnorantGuru** [comment]
@humpity
As I said in #444, "Dividers between panels are manually adjusted and don't move when the window is resized."  That has been that way since I think 0.9.1 and I don't recall anything in this behavior changing in the latest.  If you think there is something actually wrong you'll need to provide detailed steps to reproduce.

As for #444, GTK behaved poorly and there were complex issues involved, so it was decided to make it manual.  If you resize the window, you may need to adjust the dividers, etc.  Then the new positions will be remember for each combination of panels.

**Comment by humpity** [comment]
>#444, "Dividers between panels are manually adjusted and don't move when the window is resized." 
Yes, I had mistakingly thought there was somekind of fix for it (v09.4-1 ubuntu) (which do actually move??).
No biggy, I just have to get used to it.

**Comment by IgnorantGuru** [comment]
Thanks to everyone for testing.  This code has been released in 1.0.0.
